components/datetime/src/pattern/item/ule.rs

```diff
@@ -133,9 +133,126 @@ impl AsULE for PatternItem {
     }
 }
 
+/// `GenericPatternItemULE` is a type optimized for efficent storing and
+/// deserialization of `DateTimeFormat` `GenericPatternItem` elements using
+/// `ZeroVec` model.
+///
+/// The serialization model packages the pattern item in three bytes.
+///
+/// The first bit is used to disriminate the item variant. If the bit is
+/// set, then the value is the `GenericPatternItem::Placeholder` variant. Otherwise,
+/// the `GenericPatternItem::Literal` is used.
+///
+/// In case the discriminant is set:
+///
+/// 1) The rest of the first byte remains unused.
+/// 2) The second byte is unused.
+/// 3) The third byte encodes the placeholder index.
+///
+/// If the discriminant is not set, the bottom three bits of the first byte,
+/// together with the next two bytes, contain all 21 bits required to encode
+/// any [`Unicode Code Point`].
+///
+/// # Diagram
+///
+/// ```text
+/// ┌───────────────┬───────────────┬───────────────┐
+/// │       u8      │       u8      │       u8      │
+/// ├─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┼─┬─┬─┬─┬─┬─┬─┬─┤
+/// ├─┴─┴─┼─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┤
+/// │     │          Unicode Code Point             │ Literal
+/// ├─┬───┴─────────────────────────┬───────────────┤
+/// │X│                             │  Placeholder  │ Placeholder
```

Comment:
Suggestion: I think the generic data model should be "key/value", rather than "placeholder".  It is more general that way.  Users don't need to use both the key and the value.